File: scala/hamming/src/main/scala/Hamming.scala
object Hamming {
  def compute(strand1: String, strand2: String): Int = {
    val sharedLength = Math.min(strand1.length, strand2.length)

    def compareNucleotide(index: Int) = if (strand1(index) == strand2(index)) 0 else 1
    def compareNucleotides(sum: Int, index: Int) = sum + compareNucleotide(index)

    (0 until sharedLength).foldLeft(0)(compareNucleotides)
  }
}

File: java/hamming/java/hamming/src/main/java/Hamming.java
public final class Hamming {

    public static Integer compute(String strand1, String strand2) {
        Integer sum = 0;

        for (Integer i = 0; i < Math.min(strand1.length(), strand2.length()); i++)
        {
            sum += strand1.charAt(i) == strand2.charAt(i) ? 0 : 1;
        }

        return sum;
    }
}